### PR TITLE
Payments-next: Client side localization using fluent/react

### PR DESCRIPTION
**Pull request by IvoJP**
POC for l10n using fluent/react on a client side component

### CODE REVIEW

**Review by bcolsson — CHANGES_REQUESTED** [review]
Since this is POC I'm expecting this PR not to be merged. Adding a reviewing as "Request changes" to avoid confusion

File: libs/payments-ui/src/l10n/en-US.ftl [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+coupon-promo-code-applied = Promo Code Applied
+coupon-submit = Apply
+coupon-remove = Remove
+coupon-error = The code you entered is invalid or expired.
+coupon-error-generic = An error occurred processing the code. Please try again.
+coupon-error-expired = The code you entered has expired.
+coupon-error-limit-reached = The code you entered has reached its limit.
+coupon-error-invalid = The code you entered is invalid.
+# $couponDurationDate (Date) - The date at which the coupon is no longer valid, and the subscription is billed the list price.
+hello = Hello, world!
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/en-US.ftl:17]: ✂️

File: libs/payments-ui/src/l10n/fr.ftl [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+coupon-error-invalid = Le code que vous avez saisi est invalide.
+# $couponDurationDate (Date) - The date at which the coupon is no longer valid, and the subscription is billed the list price.
+coupon-enter-code =
+    .placeholder = Saisissez le code
+coupon-promo-code = Code promo
+coupon-expired = Il semble que ce code promotionnel a expiré.
+coupon-success = Votre forfait sera automatiquement renouvelé au prix courant.
+# $couponDurationDate (Date) - The date at which the coupon is no longer valid, and the subscription is billed the list price.
+coupon-success-repeating = Votre forfait sera automatiquement renouvelé au bout de { $couponDurationDate } au prix courant.
+hello = Salut le monde !
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/fr.ftl:17]: ✂️

File: libs/payments-ui/src/l10n/l10n.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
+import React, { Children, useEffect, useState, ReactNode } from 'react';
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/l10n.tsx:1]: nit: add copyright header

File: libs/payments-ui/src/l10n/en-US.ftl
Comment (pdehaan) [inline libs/payments-ui/src/l10n/en-US.ftl:1]: nit: I think most other .ftl files are just named "en.ftl"

```sh
git ls-files *.ftl | grep -Eio "\/[a-z]+\.ftl$" | sort | uniq

/branding.ftl
/en.ftl
/server.ftl
/test.ftl
```

File: libs/payments-ui/src/l10n/l10n.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+const DEFAULT_LOCALE = 'en-US';
+const AVAILABLE_LOCALES = {
+  'en-US': 'English',
+  fr: 'French',
+};
+
+async function fetchMessages(locale: string): Promise<[string, string]> {
+  const response = await fetch(String(ftl[locale]));
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/l10n.tsx:19]: cast to String feels a bit odd. Not sure if we want to use the `URL#href` prop here (or on L8-9 above) if we want a string version of the `URL` instance:
```diff
-  'en-US': new URL('./en-US.ftl', import.meta.url),
+  'en-US': new URL('./en-US.ftl', import.meta.url).href,
-  fr: new URL('./fr.ftl', import.meta.url),
+  fr: new URL('./fr.ftl', import.meta.url).href,
```

File: libs/payments-ui/src/l10n/l10n.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    const resource = new FluentResource(messages);
+    const bundle = new FluentBundle(locale);
+    bundle.addResource(resource);
+    yield bundle;
+  }
+}
+
+interface AppLocalizationProviderProps {
+  children: ReactNode;
+}
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/l10n.tsx:35]: nit: not sure if we have a style guide for this, but&hellip; should we define interfaces+types near the top of the file?
I'm guessing they aren't hoisted like functions, but felt a bit weird to be sandwiched between two function definitions. 🥪 
Although I get why maybe it makes sense to keep the interface definition closer to the implementation (L37). 🤷

File: libs/payments-ui/src/l10n/l10n.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    const fetchedMessages = await Promise.all(
+      currentLocales.map(fetchMessages)
+    );
+
+    const bundles = lazilyParsedBundles(fetchedMessages);
+    setL10n(new ReactLocalization(bundles));
+  }
+
+  if (l10n === null) {
+    return <div>Loading…</div>;
```
Comment (pdehaan) [inline libs/payments-ui/src/l10n/l10n.tsx:62]: Ah, I was going to ask if this should be localized, but I guess this has to be hardcoded to English if `l10n` is falsy&hellip;

File: libs/payments-ui/src/lib/coupon-form/coupon-form.tsx [hunk 7]
```diff
@@ -90,16 +92,18 @@ function WithoutCoupon({
...
-          ref={couponInputRef}
-        />
+        <Localized attrs={{ placeholder: true }} id="coupon-enter-code">
+          <input
+            className="coupon-input"
+            type="text"
+            name="coupon"
+            data-testid="coupon-input"
+            defaultValue={promotionCode}
+            placeholder="Enter code"
```
Comment (pdehaan) [inline libs/payments-ui/src/lib/coupon-form/coupon-form.tsx:102]: Q: "Enter code" or "Enter Code"?

https://github.com/mozilla/fxa/blob/002f0f6d2224ba945898488f635656b0253e009c/libs/payments-ui/src/l10n/en-US.ftl#L2-L3

File: libs/payments-ui/src/lib/coupon-form/coupon-form.tsx [hunk 8]
```diff
@@ -90,16 +92,18 @@ function WithoutCoupon({
...
+            disabled={disabled}
+            ref={couponInputRef}
```
Comment (pdehaan) [inline libs/payments-ui/src/lib/coupon-form/coupon-form.tsx:104]: nit: SHould this have a [`maxlength`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/maxlength)?

**Pull request closed without merging**